dipy/reconst/dki.py

```diff
@@ -889,6 +922,243 @@ def axial_kurtosis(dki_params, min_kurtosis=0, max_kurtosis=3):
+
+    Parameters
+    ----------
+    ang : array (2,)
+        arrays containing the two polar angles
+    dt : array (6,)
+        elements of the diffusion tensor of the voxel.
+    MD : float
+        mean diffusivity of the voxel
+    kt : array (15,)
+        elements of the kurtosis tensor of the voxel.
+
+    Returns
+    -------
+    neg_kt : float
+        The inverse values of the directional kurtosis for the given direction.
+
+    Notes
+    -----
+    This function is used to refine the kurtosis maxima estimate
+    
+    See also
+    --------
+    dipy.reconst.dki.kurtosis_maximum
+    """
+    n = np.array([sphere2cart(1, ang[0], ang[1])])
+    return -1. * _directional_kurtosis(dt, MD, kt, n)
+
+
+def kurtosis_maximum(dt, MD, kt, sphere, gtol=1e-5):
+    """ Computes the maxima value of a single voxel kurtosis tensor
+
+    Parameters
+    ----------
+    dt : array (6,)
+        elements of the diffusion tensor of the voxel.
+    MD : float
+        mean diffusivity of the voxel
+    kt : array (15,)
+        elements of the kurtosis tensor of the voxel.
+    sphere : Sphere class instance, optional
+        The sphere providing sample directions for the initial search of the
+        maxima value of kurtosis.
+    gtol : float, optional
+        This input is to refine kurtosis maxima under the precision of the
+        directions sampled on the sphere class instance. The gradient of the
+        convergence procedure must be less than gtol before successful
+        termination. If gtol is None, fiber direction is directly taken from 
+        the initial sampled directions of the given sphere object
+    
+    Returns
+    --------
+    max_value : float
+        kurtosis tensor maxima value
+    max_dir : array (3,)
+        Cartesian coordinates of the direction of the maximal kurtosis value 
+    """
+    # Estimation of maxima kurtosis candidates
+    AKC = _directional_kurtosis(dt, MD, kt, sphere.vertices)
+    max_val, ind = local_maxima(AKC, sphere.edges)
+    n = len(max_val)
+    max_dir = sphere.vertices[ind]
+
+    # Select the maxima from the candidates
+    max_value = max(max_val)
+    max_direction = max_dir[np.argmax(max_val.argmax)]
+
+    # refine maxima direction
+    if gtol is not None:
+        for p in range(n):
+            r, theta, phi = cart2sphere(max_dir[p, 0], max_dir[p, 1],
+                                        max_dir[p, 2])
+            ang = np.array([theta, phi])
+            ang[:] = opt.fmin_bfgs(_kt_maxima_converge, ang, args=(dt, MD, kt),
+                                   gtol=gtol, disp=False, retall=False)
+            k_dir = np.array([sphere2cart(1., ang[0], ang[1])])
+            k_val = _directional_kurtosis(dt, MD, kt, k_dir)
+            if k_val > max_value:
+                max_value = k_val
+                max_direction = k_dir
+
+    return max_value, max_direction
+
+
+def axonal_water_fraction(dki_params, sphere, mask=None, gtol=1e-5):
```

Comment:
Hi @jhlegarreta - sorry for my late reply - I have been a little bit tied with work.

Yup, the objective is really that, improve the axonal water fraction when the Da,i=0 is not met. I agree that this should be an optional parameter of the axonal_water_fraction, and also agree that you can start implementing this in a helper function that could be called by the other functions. Call it however something like axonal_water_fraction_correction - I am afraid that no diffusion MRI technique is able to give exact axonal water fractions =P.

The diffusion tensor evals are already the diffusion coefficients along axis directions of the reference frame. To get the kurtosis one, you can just use Wrotate_element(kt, indi, indj, indk, indl, B), where kt is the elements of the kurtosis tensor, B is the eigenvalues of the diffusion tensor, indi=indj=indk=indl=1 to obtain K1, indi=indj=indk=indl=2 to obtain K2 and indi=indj=indk=indl=3 to obtain K3.

The original kurtosis elements are indeed saved in dki_params together with the diffusion tensor parameters. To split diffusion and kurtois parameters use split_dki_param(dki_params).

I am afraid that to compute Da,min you have to follow a different approach than kurtosis_maxima since you need Da,min from a specified ROI while kurtosis_mixima finds the lower direction inside the voxel.